Commit various enhancements

dcalve opened the pull request:
I've been making a few changes to the cdftools as required for an update to ocean validation software. In summary, these are:

#### Enhancements
* Write time coordinate variables as double precision
  * Several supporting functions have been added in `cdfio`
  * This has **only been applied to `cdfmean` and `cdfmoc`** at the moment
* Added basin mask support and output file specification to `cdfmean`
  * New syntax: `cdfmean ... -out_suff output_file_suffix -basins n_basins basin_names`

#### Bug fixes
* Fix bug with `getvar` in `cdfio` when called for mesh mask files
  * The `getvar` function calls `getdim(...,cn_x)`, which does not work when accessing mesh mask variables
  * Since mesh mask dimension names are still hard-coded, I have added a get_dim(..., 'x') if the above fails
* Fix `cn_v...msk` namelist variables not being referenced in `cdfmoc`
* Add missing namelists and namelist variables in `modcdfnames`

Would these have value to the cdftools community? 

cc @TimGraham1982 for interest

dcalve committed:
Commit various enhancements

pmathiot commented:
Yes I think this could be useful as long as this is not mandatory. Some other routine could be concerned by your change but you don't need to do it now, we will do it when we need it.

pmathiot reviewed (commented):
I am wondering why we need to change the time variable routine (getvar1d) from REAL(4) to REAL(8).

dcalve commented:
There is truncation of double precision variables read from a file if `REAL(4)` is used in `getvar1d`. XIOS always outputs time coordinates in double precision. It is optional behaviour in the function, however.

pmathiot commented on src/modcdfnames.f90:
```diff
@@ -276,6 +276,7 @@ SUBROUTINE ReadCdfNames ()
     READ(inam, namvars    )
     READ(inam, nambathy   )
     READ(inam, namsqdvar  )
+    READ(inam, namcubvar  )
```
Is this something needed for your change or a bug you found when you worked on it ?

dcalve replied:
That's something I noticed while looking at the code

pmathiot commented on src/modcdfnames.f90:
```diff
@@ -190,8 +190,8 @@ MODULE modCdfNames
-    NAMELIST/namvars/ cn_zomsfatl, cn_zomsfglo, cn_zomsfpac, cn_zomsfinp, cn_zomsfind
-    NAMELIST/namvars/ cn_zoisoatl, cn_zoisoglo, cn_zoisopac, cn_zoisoinp, cn_zoisoind
+    NAMELIST/namvars/ cn_zomsfatl, cn_zomsfglo, cn_zomsfpac, cn_zomsfinp, cn_zomsfind, cn_zomsfinp0
+    NAMELIST/namvars/ cn_zoisoatl, cn_zoisoglo, cn_zoisopac, cn_zoisoinp, cn_zoisoind, cn_zoisoinp0
```
Well spotted

pmathiot commented on src/cdfmean.f90:
```diff
@@ -93,7 +101,8 @@ PROGRAM cdfmean
      PRINT *,' usage : cdfmean  IN-file IN-var T|U|V|F|W [imin imax jmin jmax kmin kmax]'
-     PRINT *,'       ... [-full] [-var] [-zeromean] '
+     PRINT *,'       ... [-full] [-var] [-zeromean] [-out_suff OUT-file-suffix]'
+     PRINT *,'       ... [-basins n_basins basin_names..]'
```
Do you mean prefix instead of suffix ? If I understand right, it is to add the path and runid (for example) before the ouput file name.

dcalve replied:
Yes I meant prefix, whoops!

pmathiot commented on src/cdfmean.f90:
```diff
@@ -115,10 +124,15 @@ PROGRAM cdfmean
      PRINT *,'                   partial steps.'
      PRINT *,'       [ -var ]  : also compute the spatial variance of cdfvar '
      PRINT *,'       [ -zeromean ] : create a file with cdfvar having a zero spatial mean.'
+     PRINT *,'       [ -out_suff OUT-file-suffix ] : specify SUFFIX (+ directory) of the output files'
```
I didn't really catch it before reading the code. I suggest something like "specify directory + prefix to the output name (OUT-file-suffix+cfout)"

pmathiot reviewed (commented):
I agree for all the change. I would like to make some test on cdfmean and cdfmoc to be sure it is working on whatever is the time format (FLOAT or DOUBLE) as well as for some other tools. If you have a repository ready for the test, could you send me the location ? If not, don't worry, I will find something to test with me eORCA025 run.

pmathiot commented on src/cdfmean.f90:
```diff
@@ -115,10 +124,15 @@ PROGRAM cdfmean
      PRINT *,'       [ -var ]  : also compute the spatial variance of cdfvar '
      PRINT *,'       [ -zeromean ] : create a file with cdfvar having a zero spatial mean.'
+     PRINT *,'       [ -out_suff OUT-file-suffix ] : specify SUFFIX (+ directory) of the output files'
+     PRINT *,'       [ -basins n_basins basin_names.. ] : specify basins in ',TRIM(cn_fbasins),' to be used'
```
It could be good to find a way to pickup all the basin in the file by default (ie without having to specify the number of variable and the variable name. Maybe something like a choice between your option [ -basins n_basins basin_names.. ] or [-basins_default] with no extra arguments means we pick the NEMO basin file with all the default variable names. It is a suggestion if you have time, not a requirement.

dcalve replied:
Does the NEMO basin file have an accepted convention in this respect? I'm not aware of an 'official' tool to produce such a file, so are "tmaskatl", "tmaskind" etc accepted standard names in the community? Are they used in NEMO?

I ask also because it may be an idea in the future to extend freely-defined basin mask specification to replace hard-coded basin mask names in some tools (e.g. `cdfmoc`). The mask itself can vary so "tmaskatl" is not in itself a standard, even if the name is.

pmathiot replied:
In the commit 7692fcd of modcdfname, I move the subasin variable name and file name to the NEMO one (like this no need to rename this file and the variable). To compute quantities over a specific domain (lets say Caspian sea for example) you can build a 3d mask of the Caspian sea with cdfmkmask and run your tool with this file. I agree for eORCA12, the most convenient could be to multiply the 2d mask instead of the 3d mask. We have to be clever enough to not kill the default behaviour (ie if subbasin there, use the default variable name).

pmathiot commented on src/cdfmean.f90:
```diff
@@ -115,10 +124,15 @@ PROGRAM cdfmean
      PRINT *,'      '
+     PRINT *,'     OPTIONAL FILES :'
+     PRINT *,'       File ',TRIM(cn_fbasins),'; if not available only the global domain is calculated for'
+     PRINT *,'      '
```
I suggest to change the comment by "File ',TRIM(cn_fbasins),' if -basins or -basins_default activated" because, if cn_bassin is missing and option -basins activates the tools stop (l176) and nothing is done.

dcalve replied:
Good point

pmathiot commented on src/cdfmean.f90:
```diff
@@ -146,7 +160,36 @@ PROGRAM cdfmean
+               STOP
+           ENDIF
+           CALL getarg (ijarg, cbasins(jbasin)) ; ijarg = ijarg + 1
+        END DO
```
I think we should set an exit status each time we call a STOP statement (most of the cdftools do not). So I suggest to change STOP by STOP 99 .

dcalve replied:
Ok

dcalve commented:
Note to myself to add 'time_instant' to the list of time variables in `copyatt`:

https://github.com/met-office-ocean/CDFTOOLS/blob/3c3b651fa39c1738ba1b64a49acc886d99a625d3/src/cdfio.F90#L223-L229

pmathiot commented on src/cdfio.F90:
```diff
@@ -1251,7 +1254,12 @@ FUNCTION  getvar (cdfile,cdvar,klev,kpi,kpj,kimin,kjmin, ktime, ldiom)
+       IF ( istatus /= 0 ) THEN
+          ipiglo=getdim(cdfile, 'x', ldexact=.true., kstatus=istatus)
+       ENDIF
+
```
Why this change is needed ? Was it because of the x_grid_T in the files some times ago ?

dcalve replied:
This change is needed for when `getvar` is used for mesh mask variables; the dimension names are still hard coded for this file.

dcalve commented:
Noticed a bug here where `ldexact` is needed:

https://github.com/met-office-ocean/CDFTOOLS/blob/3c3b651fa39c1738ba1b64a49acc886d99a625d3/src/cdfmean.f90#L221

In fact I think `ldexact` should be the default behaviour for `getdim` since in most cases the dimension name is read from the namelist. 

Raised #3 for this since it will apply more widely than the functions in this pull request.

dcalve committed:
Change getdim default behaviour and other fixes / changes

dcalve commented:
Updated branch with review comment changes (209c851):

**Various**
* Add temporary non-zero stop codes

**cdfio**
* Add `time_instant` to list of dimensions in `copyatt`
* Changed default behaviour of `getdim` to `ldexact = true` (#3)
  * Will require a further review of all tools to see where `ldexact = false` is still required
* Fix `kstatus` implementation for `ldexact = true` in `getdim`

**cdfmean**
* Correct file prefix description and variable name
* Correct `cn_fbasins` description

*I haven't yet included a default option for `-basins` but will do so at some point*

dcalve committed:
Remove stop codes (now in #4)

dcalve committed:
Remove missed STOP codes

dcalve commented:
Updated branch with further changes (203c2322ccaa7066e9496706ecc7d953f4533585, 79c7c077669b3d37bacd52300acb124c44666d5d).

* `cdfmean` and `cdfmoc` have now been tested with ORCA2 data
* #3 can probably be closed as I can't see any case where a partial match would be needed
  * I have left `ldexact` as a keyword for `getdim` as it may be useful

Commit:
improve standard output (set TRIM(cbas) instead of cbas)

pmathiot commented:
I relooked all my comments, your last update and I did some test wit cdfmean. After my commit, we only need to find a good solution for the dimension issue.

dcalve committed:
First attempt at partial matching for #3

dcalve committed:
Merge branch 'master' into ocean_assess_changes

dcalve committed:
Merge in changes from master

dcalve commented:
Added a first attempt at the dimension name matching issue (ab8637911713ca6097c65402ef74c52e9f13a4f1, see #3 )

dcalve commented:
The results of `cdfmean` and `cdfmoc` have not changed with the exception of `-zeromean` output for `cdfmean`, which outputs zeroes in place of the fill_value now. This was unintentional, but I don't think it's a big issue as this is closer to the behaviour of the rest of the cdftools. The `_FillValue` attribute is what needs to be changed but is not really essential.

I haven't tested the `-rapid` or `-decomp` options of `cdfmoc` as there's no reason for the `cdfmoc` results to have changed at all.

I will go ahead and merge these changes if you are happy @pmathiot.

pmathiot commented:
I agree your changes, you can merge it back to the master.
